janus_server/src/datastore.rs

```diff
@@ -1239,6 +1239,119 @@ impl<C: Clock> Transaction<'_, C> {
+        -- Join on aggregation jobs with matching task ID and aggregation parameter
+        INNER JOIN aggregation_jobs
+            ON collect_jobs.aggregation_param = aggregation_jobs.aggregation_param
+            AND collect_jobs.task_id = aggregation_jobs.task_id
+        -- Join on report aggregations with matching aggregation job ID
+        INNER JOIN report_aggregations
+            ON report_aggregations.aggregation_job_id = aggregation_jobs.id
+        -- Join on reports whose nonce falls within the collect job batch interval and which are
+        -- included in an aggregation job
+        INNER JOIN client_reports
+            ON client_reports.id = report_aggregations.client_report_id
+            AND client_reports.nonce_time <@ tsrange(
+                collect_jobs.batch_interval_start,
+                collect_jobs.batch_interval_start
+                    + collect_jobs.batch_interval_duration * interval '1 second')
+        WHERE
+            -- Constraint for tasks table in FROM position
+            tasks.id = collect_jobs.task_id
+            -- Do not acquire collect jobs that have been completed
+            AND collect_jobs.helper_aggregate_share IS NULL
+            -- Do not acquire collect jobs with an unexpired lease
+            AND collect_jobs.lease_expiry <= $2
+        GROUP BY collect_jobs.id
+        -- Do not acquire collect jobs where any associated aggregation jobs are not finished
+        HAVING bool_and(aggregation_jobs.state != 'IN_PROGRESS')
+        -- Honor maximum_acquire_count *after* winnowing down to runnable collect jobs
+        LIMIT $3
+    )
+    RETURNING tasks.task_id, tasks.vdaf, collect_jobs.collect_job_id, collect_jobs.id
+)
+SELECT task_id, vdaf, collect_job_id FROM updated
+-- TODO (issue #174): revisit collect job queueing behavior implied by this ORDER BY
+ORDER BY id DESC
+"#,
+            )
+            .await?;
+        self.tx
+            .query(
+                &stmt,
+                &[
+                    /* lease_expiry */ &lease_expiry_time.as_naive_date_time(),
+                    /* now */ &now.as_naive_date_time(),
+                    /* limit */ &maximum_acquire_count,
+                ],
+            )
+            .await?
+            .into_iter()
+            .map(|row| {
+                let task_id = TaskId::get_decoded(row.get("task_id"))?;
+                let vdaf = row.try_get::<_, Json<VdafInstance>>("vdaf")?.0;
+                let collect_job_id = row.get("collect_job_id");
+                Ok((
+                    AcquiredCollectJob {
+                        task_id,
+                        vdaf,
+                        collect_job_id,
+                    },
+                    lease_expiry_time,
+                ))
+            })
+            .collect()
+    }
+
+    /// release_collect_job releases an acquired (via e.g. acquire_incomplete_collect_jobs) collect
+    /// job. It returns an error if the collect job has no current lease.
```

Comment:
This method doesn't currently check for an existing lease. Should this be an IOU for when we add ticket IDs to both of our kinds of leases?

Reply:
I spotted Brandon's comment in the corresponding `release_aggregation_job`. I don't think it makes sense to check for an active lease, because then the caller is racing with the lease expiring. Whether the lease is released due to explicit caller action or due to timeout, the desired postcondition (no lease is held) is achieved, so I don't think it makes sense to hand an error to the caller (what would they do with it?).

Reply:
Oh, wait, I realize now you must both mean checking that the lease is not `-infinity`, rather than checking if there is an unexpired lease. That seems like a sensible guard against programmer error. I think this fits with a few other hardenings and improvements of our queue implementation we'll need that I think should be addressed beyond this PR. I have filed #193 as an umbrella issue to track that work.

Reply:
Yep, avoiding the ABA problem in the presence of significant clock skew (or implementation bugs in lease timeout logic) is the goal of that TODO. I think it's worthwhile to address because, while clock skew may not be very likely, the consequences of the problem if it does occur are effectively "data corruption" or "lost writes".

Reply:
Got it. I'm not arguing that this isn't a problem, but rather that we should address it in a later PR, fixing this for both aggregate jobs and collect jobs at once.